From 19763701bf5938fef4583b99cc4b040155ec3882 Mon Sep 17 00:00:00 2001
From: Mikolaj Staworzynski <mikolaj.staworzynski@redembedded.com>
Date: Wed, 6 Mar 2024 14:57:10 +0100
Subject: [PATCH] ONEM-34263: do not unload libWPEFrameworkWebKitBrowserImpl.so

There is a problem with unloading that library, library has dependency on libWPEWebkit
and libWPEWebkit contains some per process static objects and threads.
Two randomly observed issues on WebKitBrowser plugin deactivation was for the
BMScavenger thread and PressureMonitor threads from libWPEWebkit.
Library is loaded by WPEProcess that handle webkitbrowser plugin not by the WPEFramework.

Examples:

That one was 100 % on gdb, not at all after that fix.
Thread 16 "BMScavenger" received signal SIGABRT, Aborted.
[Switching to Thread 20771.20791]
__libc_do_syscall () at libc-do-syscall.S:49
49	libc-do-syscall.S: No such file or directory.
Breakpoint 1 at 0xb61f1700: file abort.c, line 54.
0  __libc_do_syscall () at libc-do-syscall.S:49
1  0xb61fe490 in __libc_signal_restore_set (set=0x8db5fa50) at ../sysdeps/unix/sysv/linux/internal-signals.h:86
2  __GI_raise (sig=sig@entry=6) at ../sysdeps/unix/sysv/linux/raise.c:48
3  0xb61f17a2 in __GI_abort () at abort.c:79
4  0xb621f8a6 in __libc_message (action=action@entry=do_abort, fmt=<optimized out>) at ../sysdeps/posix/libc_fatal.c:155
5  0xb621f928 in __GI___libc_fatal (message=0xb644d128 "The futex facility returned an unexpected error code.\n") at ../sysdeps/posix/libc_fatal.c:164
6  0xb64486f8 in futex_fatal_error () at ../sysdeps/nptl/futex-internal.h:200
7  futex_wait_cancelable (private=<optimized out>, expected=<optimized out>, futex_word=<optimized out>) at ../sysdeps/nptl/futex-internal.h:200
8  __pthread_cond_wait_common (abstime=0x0, clockid=0, mutex=<optimized out>, cond=0xb629a422) at pthread_cond_wait.c:508
9  __pthread_cond_wait (cond=0xb629a422, mutex=0x2d6ed7c4) at pthread_cond_wait.c:638
10 0xb636b476 in _gthread_cond_wait (_mutex=<optimized out>, __cond=<optimized out>) at /usr/src/debug/gcc-runtime/9.3.0-r0/arm-rdk-linux-gnueabi/libstdc++-v3/include/arm-rdk-linux-gnueabi/bits/gthr-default.h:865
11 std::condition_variable::wait (this=<optimized out>, __lock=...) at ../../../../../../../../../../work-shared/gcc-9.3.0-r0/gcc-9.3.0/libstdc++-v3/src/c++11/condition_variable.cc:53
12 0xb540ed72 in ?? ()

That one catched one time on gdb.
Thread 17 "PressureMonitor" received signal SIGSEGV, Segmentation fault.
[Switching to Thread 19108.19129]
0xb34ce7dc in strcmp@plt () from ./rootfs/appmodule/wpe/usr/lib/libWPEWebKit-1.0.so.3
Breakpoint 1 at 0xb3678d78: abort. (25 locations)
0  0xb34ce7dc in strcmp@plt () from ./rootfs/appmodule/wpe/usr/lib/libWPEWebKit-1.0.so.3
1  0xb386121c in systemMemoryUsedAsPercentage () at /usr/src/debug/wpe-webkit/1_2.38+git-r0/git/Source/WebKit/UIProcess/linux/MemoryPressureMonitor.cpp:260
2  operator() () at /usr/src/debug/wpe-webkit/1_2.38+git-r0/git/Source/WebKit/UIProcess/linux/MemoryPressureMonitor.cpp:372
---
 Source/core/Library.cpp | 6 +++++-
 1 file changed, 5 insertions(+), 1 deletion(-)

diff --git a/Source/core/Library.cpp b/Source/core/Library.cpp
index eb294b7..cd6bcc6 100644
--- a/Source/core/Library.cpp
+++ b/Source/core/Library.cpp
@@ -180,7 +180,11 @@ namespace Core {
                 }
 
 #ifdef __LINUX__
-                dlclose(_refCountedHandle->_handle);
+                // there is a problem with unloading that library, library has dependency on libWPEWebkit
+                // and libWPEWebkit contains some per process static objects and threads
+                // two randomly observed issues was for the BMScavenger thread and PressureMonitor threads from there
+                if (_refCountedHandle->_name.find("libWPEFrameworkWebKitBrowserImpl.so") == std::string::npos)
+                    dlclose(_refCountedHandle->_handle);
 #endif
 #ifdef __WINDOWS__
                 ::FreeLibrary(_refCountedHandle->_handle);
-- 
2.25.1

